google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerOptions.java

```diff
@@ -433,7 +433,41 @@ public TransportOptions getDefaultTransportOptions() {
   }
 
   protected SpannerRpc getSpannerRpcV1() {
-    return (SpannerRpc) getRpc();
+    return getRpc();
+  }
+
+  /**
+   * Returns a {@link Spanner} service object. This method will create a {@link Spanner} instance on
+   * the first call and subsequent calls to this method will return the same {@link Spanner}
+   * instance. If the instance is closed, the following call to this method will create a new {@link
+   * Spanner} instance.
+   */
+  @Override
+  public Spanner getService() {
+    // Method is only overridden in order to supply additional documentation.
```

Comment:
My preference here would be to instead add documentation to getService in ServiceOptions so we don't have to override this (same with getRpc).

Reply:
That's not possible as there are Service classes other than Spanner that are not closeable and do not have an `isClosed()` method, which means that the behavior that is described for Spanner is only applicable to Spanner. That is probably also the reason that the default implementation is to cache the Service object indefinitely without checking its validity.

One possible generic solution could be to change the `ServiceOptions` class to implement the behavior of this PR for all Service classes that implement a custom interface `CloseableService`. All other Service classes would keep the current default behavior.

Reply:
Oh, sorry, didn't realize that was part of google-cloud-core. Can we perhaps add documentation to shouldRefreshService?

Reply:
I've added documentation to the shouldRefresh... methods.
Note that we cannot/should not rely on the documentation of these methods for our users, as these methods are protected.